server/src/main/java/org/elasticsearch/transport/ConnectionManager.java

```diff
@@ -122,40 +118,37 @@ public void connectToNode(DiscoveryNode node, ConnectionProfile connectionProfil
             return;
         }
 
-        try (Releasable lock = connectionLock.acquire(node.getId())) {
-            Transport.Connection connection = connectedNodes.get(node);
-            if (connection != null) {
-                assert connectingNodes.containsKey(node) == false;
-                lock.close();
-                connectingRefCounter.decRef();
-                listener.onResponse(null);
-                return;
-            }
+        if (pendingConnections.containsKey(node)) {
```

Comment:
I'm confused. If we have a pending connection, a second concurrent attempt now returns "OK", making it look like a connection has already been established?

Reply:
It also looks like we need more tests around this.

Reply:
I fixed this bug and modified the concurrency test to ensure that it will occasionally fail if this is broken again.